crates/starknet-devnet-core/src/state/mod.rs

```diff
@@ -55,39 +67,46 @@ pub trait CustomState {
 }
 
 #[derive(Default, Clone)]
-/// Utility structure that makes it easier to calculate state diff later on
+/// Utility structure that makes it easier to calculate state diff later on. Classes are first
+/// inserted into the staging area (pending state), later to be committed (latest state).
 pub struct CommittedClassStorage {
     staging: HashMap<ClassHash, ContractClass>,
-    committed: HashMap<ClassHash, ContractClass>,
+    committed: HashMap<ClassHash, (ContractClass, u64)>,
 }
 
 impl CommittedClassStorage {
     pub fn insert(&mut self, class_hash: ClassHash, contract_class: ContractClass) {
         self.staging.insert(class_hash, contract_class);
     }
 
-    pub fn commit(&mut self) -> HashMap<ClassHash, ContractClass> {
+    pub fn commit(&mut self, block_number: u64) -> HashMap<ClassHash, ContractClass> {
         let diff = self.staging.clone();
-        self.committed.extend(self.staging.drain());
+        let numbered =
+            self.staging.drain().map(|(class_hash, class)| (class_hash, (class, block_number)));
+        self.committed.extend(numbered);
         diff
     }
 }
 
 pub struct StarknetState {
     pub(crate) state: CachedState<DictState>,
-    rpc_contract_classes: CommittedClassStorage,
+    /// The class storage is meant to be shared between states to prevent copying (due to memory
+    /// concerns). Knowing which class was added when is made possible by storing the class
+    /// together with the block number.
```

Comment:
Why do we use block numbers instead of block hashes?

Reply:
Answered in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/499#discussion_r1650838115